File: reacjilator-scala-jslack/src/main/scala/slackapp/handler/WarmupHandler.scala
package slackapp.handler

import com.amazonaws.services.lambda.model.InvokeRequest
import com.amazonaws.services.lambda.runtime.{ Context, RequestHandler }
import com.github.seratch.jslack.app_backend.vendor.aws.lambda.request.ApiGatewayRequest
import com.github.seratch.jslack.app_backend.vendor.aws.lambda.response.ApiGatewayResponse
import com.github.seratch.jslack.common.json.GsonFactory
import org.slf4j.LoggerFactory
import slackapp.service.AmazonWebServices

import scala.collection.JavaConverters._

/**
  * To address cold start problems
  */
class WarmupHandler extends RequestHandler[ApiGatewayRequest, ApiGatewayResponse] {
  import WarmupHandler._

  override def handleRequest(req: ApiGatewayRequest, context: Context): ApiGatewayResponse = {
    log.info(s"request: ${req}")

    val lambda = aws.newLambdaClient
    val functionNames: Seq[String] = {
      val serverlessStage = aws.serverlessStage
      Seq(
        s"reacjilator-${serverlessStage}-echo",
        s"reacjilator-${serverlessStage}-events"
      )
    }
    for (functionName <- functionNames) {
      val result = lambda.invoke(invokeRequest(functionName, payloadString))
      if (result.getStatusCode != 200) {
        log.error(s"Failed to warmup the function: ${functionName}, result: ${result.getFunctionError}")
      }
    }
    response
      .setStatusCode(200)
      .setObjectBody("Done")
      .build
  }

}

object WarmupHandler {

  val payload       = Map("warmup" -> true)
  val payloadString = GsonFactory.createSnakeCase.toJson(payload.asJava)

  private val log = LoggerFactory.getLogger(classOf[WarmupHandler])
  private val aws = new AmazonWebServices

  def invokeRequest(functionName: String, payload: String) = {
    new InvokeRequest()
      .withFunctionName(functionName)
      .withPayload(payload)
  }

  def response = ApiGatewayResponse.builder

}

File: reacjilator-java-sdk/src/main/java/slackapp_backend/handler/WarmupHandler.java
package slackapp_backend.handler;

import com.amazonaws.services.lambda.AWSLambda;
import com.amazonaws.services.lambda.model.InvokeRequest;
import com.amazonaws.services.lambda.model.InvokeResult;
import com.amazonaws.services.lambda.runtime.Context;
import com.amazonaws.services.lambda.runtime.RequestHandler;
import com.slack.api.bolt.aws_lambda.request.ApiGatewayRequest;
import com.slack.api.bolt.aws_lambda.response.ApiGatewayResponse;
import com.slack.api.util.json.GsonFactory;
import lombok.extern.slf4j.Slf4j;
import slackapp_backend.service.AmazonWebServices;

import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * To address cold start problems
 */
@Slf4j
public class WarmupHandler implements RequestHandler<ApiGatewayRequest, ApiGatewayResponse> {

    private static final AmazonWebServices AWS = new AmazonWebServices();

    private static final Map<String, Object> PAYLOAD;
    public static final String PAYLOAD_STRING;

    static {
        PAYLOAD = new HashMap<>();
        PAYLOAD.put("warmup", true);
        PAYLOAD_STRING = GsonFactory.createSnakeCase().toJson(PAYLOAD);
    }

    @Override
    public ApiGatewayResponse handleRequest(ApiGatewayRequest req, Context context) {
        log.info("request: {}", req);

        AWSLambda lambda = AWS.createLambdaClient();
        String serverlessStage = AWS.getServerlessStage();
        List<String> functionNames = Arrays.asList(
                "reacjilator-" + serverlessStage + "-echo",
                "reacjilator-" + serverlessStage + "-events"
        );
        for (String functionName : functionNames) {
            InvokeRequest invokeReq = new InvokeRequest().withFunctionName(functionName).withPayload(PAYLOAD_STRING);
            InvokeResult result = lambda.invoke(invokeReq);
            if (result.getStatusCode() != 200) {
                log.error("Failed to warmup the function: {}, result: {}", functionName, result.getFunctionError());
            }
        }

        return ApiGatewayResponse.builder()
                .statusCode(200)
                .objectBody("Done")
                .build();
    }
}